homeassistant/components/zha/entity.py

```diff
@@ -95,265 +81,67 @@ def device_info(self) -> DeviceInfo:
             return
 
-        self._handled_group_membership = True
-        await self.async_remove(force_remove=True)
-        if len(self._group.members) >= 2:
-            async_dispatcher_send(
-                self.hass, SIGNAL_GROUP_ENTITY_REMOVED, self._group_id
-            )
-
-    async def async_added_to_hass(self) -> None:
-        """Register callbacks."""
-        await super().async_added_to_hass()
-        await self.async_update()
-
-        self.async_accept_signal(
-            None,
-            f"{SIGNAL_GROUP_MEMBERSHIP_CHANGE}_0x{self._group_id:04x}",
-            self._handle_group_membership_changed,
-            signal_override=True,
-        )
-
-        if self._change_listener_debouncer is None:
-            self._change_listener_debouncer = Debouncer(
-                self.hass,
-                _LOGGER,
-                cooldown=self._update_group_from_child_delay,
-                immediate=False,
-                function=functools.partial(self.async_update_ha_state, True),
-            )
-            self.async_on_remove(self._change_listener_debouncer.async_cancel)
-        self._async_unsub_state_changed = async_track_state_change_event(
-            self.hass, self._entity_ids, self.async_state_changed_listener
-        )
+        self.restore_external_state_attributes(state)
 
     @callback
-    def async_state_changed_listener(self, event: Event[EventStateChangedData]) -> None:
-        """Handle child updates."""
-        # Delay to ensure that we get updates from all members before updating the group
-        assert self._change_listener_debouncer
-        self._change_listener_debouncer.async_schedule_call()
+    def restore_external_state_attributes(self, state: State) -> None:
+        """Restore ephemeral external state from Home Assistant back into ZHA."""
+
+        # Some operations rely on extra state that is not maintained in the ZCL
+        # attribute cache. Until ZHA is able to maintain its own persistent state (or
+        # provides a more generic hook to utilize HA to do this), we directly restore
+        # them.
 
     async def async_will_remove_from_hass(self) -> None:
-        """Handle removal from Home Assistant."""
+        """Disconnect entity object when removed."""
+        for unsub in self._unsubs[:]:
+            unsub()
+            self._unsubs.remove(unsub)
         await super().async_will_remove_from_hass()
-        if self._async_unsub_state_changed is not None:
-            self._async_unsub_state_changed()
-            self._async_unsub_state_changed = None
+        self.remove_future.set_result(True)
 
+    @convert_zha_error_to_ha_error
     async def async_update(self) -> None:
-        """Update the state of the group entity."""
+        """Update the entity."""
+        await self.entity_data.entity.async_update()
+        self.async_write_ha_state()
+
+    def log(self, level: int, msg: str, *args, **kwargs):
```

Comment:
```suggestion
    def log(self, level: int, msg: str, *args, **kwargs) -> None:
```
also, isn't level an enum?

Reply:
Nope the levels are constants:
```python
CRITICAL = 50
FATAL = CRITICAL
ERROR = 40
WARNING = 30
WARN = WARNING
INFO = 20
DEBUG = 10
NOTSET = 0
```